### PR TITLE
Recoil calibration lowPU, support for lowPU 2023 samples

### PR DESCRIPTION
**Pull request by jeyserma**
- Fixed electrons for lowPU (data/MC agree much better now)
- More reasonable binning for the fake rate estimate in lowPU
- Derived recoil calibration for lowPU RawPFMET for both muon and electron channels, including statistical uncertainties
- Added support to read 2023 lowPU samples
    - ME (mass, width, sin2theta) weights read from a different branch w.r.t old samples, reorganized that code
    - Treat the lowPU samples as different ERAs, using NanoAOD v12
    - Involved rearranging of the data directories for lowPU (both at CERN and SubMIT)

This PR should not touch the central value nor the uncertainties of the highPU.

### CODE REVIEW

**Comment by jeyserma** [comment]
No changes in the highPU as expected. LowPU changes also as expected (fix of the electron channel + recoil calibration).

Can be merged from my side.

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -281,27 +281,27 @@ jobs:
...
         run: >- 
           scripts/ci/run_with_singularity.sh scripts/ci/setup_and_run_python.sh scripts/plotting/makeDataMCStackPlot.py --fakeEstimation simple --fakeSmoothingMode binned
-          --yscale 1.5 --baseName lep_pt_eta_phi --hists eta phi -p mu -o $WEB_DIR -f $PLOT_DIR/lowPU $HIST_FILE_MU
+          --forceGlobalScaleFakes 1.0 --yscale 1.5 --baseName lep_pt_eta_phi --hists eta phi -p mu -o $WEB_DIR -f $PLOT_DIR/lowPU $HIST_FILE_MU
 
       - name: lowpu w mu plot mt
         run: >- 
           scripts/ci/run_with_singularity.sh scripts/ci/setup_and_run_python.sh scripts/plotting/makeDataMCStackPlot.py --fakeEstimation simple --fakeSmoothingMode binned
-          --rebin 2 --baseName transverseMass --hists mt -p mu -o $WEB_DIR -f $PLOT_DIR/lowPU $HIST_FILE_MU
+          --forceGlobalScaleFakes 1.0 --rebin 2 --baseName transverseMass --hists mt -p mu -o $WEB_DIR -f $PLOT_DIR/lowPU $HIST_FILE_MU
```
Comment (kdlong) [inline .github/workflows/main.yml:304]: This means the fake norm needs to be scaled by 1.0 for muons and 1.1 6 for electrons?

File: scripts/histmakers/mw_lowPU.py [hunk 2]
```diff
@@ -10,11 +10,16 @@
 parser.add_argument("--noGenMatchMC", action='store_true', help="Don't use gen match filter for prompt muons with MC samples (note: QCD MC never has it anyway)")
 parser.add_argument("--flavor", type=str, choices=["e", "mu"], help="Flavor (e or mu)", default="mu")
 
+parser = common.set_parser_default(parser, "pt", [34, 25, 1000])
```
Comment (kdlong) [inline scripts/histmakers/mw_lowPU.py:13]: Do you really want uniform binning up to 1000?

File: scripts/histmakers/mw_lowPU.py [hunk 3]
```diff
@@ -37,27 +42,33 @@
...
-mtw_min=40 # for Wmass (roughly half the boson mass)
+mtw_min = 40 # for Wmass (roughly half the boson mass)
+
+# lepton cuts
+lep_pt_min = 25
 
 # axes used in fakerate calculation
-axis_fakes_pt = hist.axis.Variable(common.get_binning_fakes_pt(args.pt[1], args.pt[2]), name = "pt", underflow=False)
-axis_fakes_eta = hist.axis.Regular(int((args.eta[2]-args.eta[1])*10/2), args.eta[1], args.eta[2], name = "eta", underflow=False, overflow=False)
+axis_fakes_pt = hist.axis.Variable([25, 30, 35, 40, 45, 50, 55], name = "pt", underflow=False, overflow=True)
```
Comment (kdlong) [inline scripts/histmakers/mw_lowPU.py:59]: I guess this will need to go higher in pt too at some point?

File: scripts/histmakers/mw_lowPU.py [hunk 4]
```diff
@@ -240,7 +234,7 @@ def build_graph(df, dataset):
             df = df.Define("prefireCorr", "wrem::prefireCorr(0, Jet_pt, Jet_eta, Jet_phi, Jet_muEF, Jet_neEmEF, Jet_chEmEF, Photon_pt, Photon_eta, Photon_phi, Lep_pt, Lep_eta, Lep_phi)")
             df = df.Define("SFMC", "lepSF_IDISO*lepSF_HLT*prefireCorr")
 
-        df = df.Define("exp_weight", "1.0")# "SFMC")
+        df = df.Define("exp_weight", "SFMC")
```
Comment (kdlong) [inline scripts/histmakers/mw_lowPU.py:237]: Is SFMC defined for electrons?

File: scripts/histmakers/mw_lowPU.py [hunk 5]
```diff
@@ -310,162 +305,24 @@ def build_graph(df, dataset):
 
             results.append(df.HistoBoost(f"{n}_luminosity", a, [*c, "luminosityScaling"], tensor_axes = [common.down_up_axis], storage=hist.storage.Double()))
 
-            if not args.noRecoil:
+            if not args.noRecoil and args.recoilUnc:
```
Comment (kdlong) [inline scripts/histmakers/mw_lowPU.py:308]: Why do we now have args.recoilUnc and args.noRecoilUnc? Can't it be one arg?

File: .github/workflows/main.yml (hunk 1 above)
Comment (jeyserma) [reply .github/workflows/main.yml:304]: Yes, the norm scale factors are quite different for the lowPU, which are obtained visually by my plotting scripts. Though the plots from the CI don't close as I expected, I will debug this later

File: scripts/histmakers/mw_lowPU.py (hunk 2 above)
Comment (jeyserma) [reply scripts/histmakers/mw_lowPU.py:13]: Also this has to be optimized: for the lowPU we don't want an upper cut on lepton pT, but currently the unfolding needs these arguments so I put a high number for now.

**Comment by kdlong** [comment]
I made some small style comments but they aren't so important

File: scripts/histmakers/mw_lowPU.py (hunk 3 above)
Comment (jeyserma) [reply scripts/histmakers/mw_lowPU.py:59]: Yes, the fakes for the lowPU are to be studied in detail

File: scripts/histmakers/mw_lowPU.py (hunk 4 above)
Comment (jeyserma) [reply scripts/histmakers/mw_lowPU.py:237]: Yes, is defined above

File: scripts/histmakers/mw_lowPU.py [hunk 6]
```diff
@@ -284,15 +280,14 @@ def build_graph(df, dataset):
...
-                        
+        for n, c, a in (("nominal", cols, axes), ("transverseMass", cols_mt, axes_mt)):
             results.append(df.HistoBoost(f"{n}_prefireCorr", [*a], [*c, "prefireCorr_syst_tensor"], tensor_axes = [common.down_up_axis]))
 
             if dataset.name in common.vprocs_lowpu:
                 df = syst_tools.add_theory_hists(results, df, args, dataset.name, corr_helpers, qcdScaleByHelicity_helper, a, c, base_name=n)
 
             # lepton efficiencies
-            if flavor == "mu":
+            if flavor == "mu" and True:
```
[the remark this answers is not in the document]
Comment (jeyserma) [reply scripts/histmakers/mw_lowPU.py:290]: It is used to have some uncertainties on the efficiencies for the electrons. I can remove it, as it was a test indeed (can remove in the next iteration?)

File: scripts/histmakers/mw_lowPU.py (hunk 5 above)
Comment (jeyserma) [reply scripts/histmakers/mw_lowPU.py:308]: We still want to have the flexibility to run without recoil corrections and/or without recoil uncertainties (especially as running with uncertainties is quite slow for the highPU)

**Comment by jeyserma** [comment]
It's true that some items require some major revision for the lowPU (fakes, uncertainties, unfolding), but we'll do that in the near term future.

**Comment by bendavid** [comment]
You'll need to update the wremnants-data commit to a31cceadbb49ea7589da85b50f993df98c2a84b8 (because of the veto PR's which were just merged)

**Comment by bendavid** [comment]
Ok github let me take care of it

**Pull request merged**